File: sections/35_math_mixed_contest/problems/n_squarefree_queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    cin >> q;
    vector<int> queries(q);
    int maximum = 1;
    for (int& x : queries) {
        cin >> x;
        maximum = max(maximum, x);
    }

    vector<char> squarefree(maximum + 1, true);
    squarefree[0] = false;
    for (int p = 2; p * p <= maximum; ++p) {
        for (int multiple = p * p; multiple <= maximum; multiple += p * p)
            squarefree[multiple] = false;
    }

    vector<int> prefix(maximum + 1);
    for (int i = 1; i <= maximum; ++i)
        prefix[i] = prefix[i - 1] + squarefree[i];

    for (int x : queries)
        cout << prefix[x] << '\n';
}

File: sections/35_math_mixed_contest/problems/n_squarefree_queries/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    queries = data[1:1 + q]
    maximum = max(queries, default=1)

    squarefree = [True] * (maximum + 1)
    squarefree[0] = False
    p = 2
    while p * p <= maximum:
        for multiple in range(p * p, maximum + 1, p * p):
            squarefree[multiple] = False
        p += 1

    prefix = [0] * (maximum + 1)
    for value in range(1, maximum + 1):
        prefix[value] = prefix[value - 1] + squarefree[value]

    print("\n".join(str(prefix[x]) for x in queries))


if __name__ == "__main__":
    main()
